spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala

```diff
@@ -309,68 +301,42 @@ object DistriOptimizer {
         // enough records were processed for this batch, so update the model
         val value = lossSum.value / numFinishedModelUpdates
 
-        var l2Norm = 0.0f
-        var scale = ev.fromType(numFinishedModelUpdates)
-        if (normClippingEnable) {
-          val sumSquare = models.mapPartitions(modelIter => {
-            val getG = System.nanoTime()
-            parameters.aggregateGradientPartition()
-            driverMetrics.add("aggregrateGradientParition average executor",
-              System.nanoTime() - getG)
-
-            val gradLength = parameters.gradientPartition.nElement()
-            val taskSize = gradLength / _subModelNumber
-            val extraTask = gradLength % _subModelNumber
-            val parallelNum = if (taskSize == 0) extraTask else _subModelNumber
-            val squares = new Array[Double](parallelNum)
-            Engine.default.invokeAndWait((0 until parallelNum).map(tid => () => {
-              val offset = tid * taskSize + math.min(tid, extraTask)
-              val length = taskSize + (if (tid < extraTask) 1 else 0)
-              squares(tid) = ev.toType[Double](
-                parameters.gradientPartition.narrow(1, offset + 1, length).sumSquare())
-            }))
-            var sum = 0.0
-            var i = 0
-            while (i < parallelNum) {
-              sum += squares(i)
-              i += 1
-            }
-            Iterator.single(sum)
-          }).reduce(_ + _)
-          l2Norm = (math.sqrt(sumSquare) / numFinishedModelUpdates).toFloat
-          if (l2Norm > normValueClip) {
-            scale = ev.fromType[Double]((l2Norm * numFinishedModelUpdates) / normValueClip)
-          }
-        }
+        driverState("numFinishedModel") = numFinishedModelUpdates
+        // aggregateG is flag to mark whether gradient is updated. May changed in the future.
+        driverState("aggregateG") = false
+        // parameterProcesser may aggregate gradient,
+        // and change the value of aggregateG in driverState.
+        parameterProcessers.foreach(_.collectGlobalData(models, parameters, metrics, driverState))
+        val aggregateG = driverState[Boolean]("aggregateG")
+        val stateBroadcast = sc.broadcast(driverState)
 
         models.mapPartitions { modelIter =>
           val modelCache = modelIter.next()
-          if (!normClippingEnable) {
+          // if parameterProcesser has aggregated gradient, we can skip this aggregation.
+          if (!aggregateG) {
             val getG = System.nanoTime()
-            parameters.aggregateGradientPartition()
+            parameters.aggregateGradientPartition(numFinishedModelUpdates)
             driverMetrics.add("aggregrateGradientParition average executor",
               System.nanoTime() - getG)
           }
-          parameters.gradientPartition.div(scale)
+          parameterProcessers.foreach(
+            _.processParameters(parameters, modelCache, stateBroadcast.value))
```

Comment:
@yiheng I make a little change here, will this help to avoid the cost of broadcast if parameterProcessers is empty?